src/main/java/graphql/annotations/DefaultTypeFunction.java

```diff
@@ -230,9 +230,11 @@ public GraphQLType apply(Class<?> aClass, AnnotatedType annotatedType) {
         @Override
         public GraphQLType apply(Class<?> aClass, AnnotatedType annotatedType) {
             GraphQLName name = aClass.getAnnotation(GraphQLName.class);
-            String typeName = name == null ? aClass.getSimpleName() : name.value();
-            if (processing.containsKey(typeName)) {
-                return processing.get(typeName);
+            String typeName = name == null ? aClass.getName() : name.value();
+            if (types.containsKey(typeName)) {
+                return types.get(typeName);
+            } else if (processing.containsKey(typeName)) {
+                return processing.getOrDefault(typeName, new GraphQLTypeReference(typeName));
```

Comment:
I think this was the original idea with "processing". The race condition provoked the thread to get into the containsKey if but retrieving null with the get inside.